File: Function/dijkstra.cpp
int dijkstra(int n, vector<pair<int,int>> &graph, int node1, int node2){
    vector<int> dist(n+1, INT_MAX);
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
    dist[node1] = 0;
    pq.push({0, node1});
    while(!pq.empty()){
        int d = pq.top().first;
        int u = pq.top().second;
        pq.pop();
        if(u == node2) return d;
        if(d > dist[u]) continue;
        for(auto &v: graph[u]){
            if(dist[u] + v.second < dist[v.first]){
                dist[v.first] = dist[u] + v.second;
                pq.push({dist[v.first], v.first});
            }
        }
    }
    return -1;
}

File: Function/dijkstra.py
from heapq import *

def dijkstra(n, graph, node1, node2):
    # n: number of nodes
    # graph: adjacency matrix
    # node1: start node
    # node2: end node
    # return: shortest path from node1 to node2
    #         and the length of the path

    # initialize
    heap = [] # heap
    heappush(heap, [0, node1]) # push the start node into the heap
    dist = [float("inf")] * n
    dist[node1] = 0

    # find the shortest path
    while heap:
        d, u = heappop(heap)
        if u == node2:
            return d
        if d > dist[u]:
            continue
        for v, w in graph[u]:
            if dist[v] > dist[u] + w:
                dist[v] = dist[u] + w
                heappush(heap, [dist[v], v])
    return -1
